### ISSUE TITLE
Atoms 'flying away' in GAP MD

### ISSUE THREAD

**Issue by dmilardovich** [libAtoms/QUIP#454]
Hi everyone, 
 
 I am currently working on a GAP for the oxidation of Si surfaces. I am currently facing the problem that some atoms 'fly away' from the structure in an unphysical way (screenshot attached). I was wondering if you could please answer a couple of questions to help me solve this problem: 

 1. How common is this problem? This is an MD at 1000K, with a time-step of 0.50 fs. 
 2. I am currently training only on energies, because of computing resources limitations. Could this problem be solved simply by training on energies AND forces? 
 3. Next are the parameters I am using. Could you please tell me if you spot anything out of place? The training dataset is composed of 1750 snaps (DFTB), containing dimmers, clusters, surfaces, snaps from melt-and-quench MDs and even snaps from Si oxidation MDs which are nearly identical to those found in the test MD in which atoms 'fly away'. 

os.system("gap_fit energy_parameter_name=energy at_file=GAP_TRAIN_dimers_specified_BIG_amorphous.xyz gap={distance_Nb order=2 cutoff=2.0 n_sparse=15 covariance_type=ard_se delta=4 theta_uniform=2.5 sparse_method=uniform compact_clusters=T : soap atom_sigma=0.5 l_max=6 n_max=6 cutoff=4.0 cutoff_transition_width=1.0 delta=0.4 covariance_type=dot_product n_sparse=1000 zeta=4} default_sigma={0.002 0.0 0.0 0.0} config_type_sigma={bulk:0.005:0.0:0.0:0.0:amorphous:0.005:0.0:0.0:0.0:cluster:0.01:0.0:0.0:0.0} e0={Si:-29.7169:O:-83.9795:H:-6.4926} sparse_jitter=1e-8 gp_file=GAP_SOAP_2B_dimmers_specified_BIG_amorphous.xml")

![GAP_MD](https://user-images.githubusercontent.com/47321981/183431929-a85bca73-5285-4962-bcf9-f486c58bbcf8.png)

Note: I clarify that I reduced the 2-body descriptor cutoff to 2.0 A, as otherwise (with 4.0 A) the MDs tended to collapse into a large unphysical cluster.  

Thanks a lot for your time! 
Best, 
Diego

**Comment by albapa** [comment]
Hi,

1. I think your potential is not quite ready for MD yet - the potential doesn't "know" that clusters are energetically unfavourable. The easiest way to fix it is feeding the unphysical configurations back into your training data (after evaluating with your ab initio method), and retrain your GAP.

Regarding the other points:
2. Forces carry incredibly rich information - you'd need a **lot** of energy-only training data to make up for it. With the new data-parallel `gap_fit` resources shouldn't be a problem if you have access to a few interconnected nodes. Check it out [here](https://arxiv.org/abs/2207.03803).
3. I don't remember what the typical bond lengths are but I don't see 2.0 cutoff doing much. Again, forces in the training data should inform the 2b model much better. Note that your `delta` for the 2b is huge - that's the variance per descriptor - and you will have a lot of those. Other have published strategies on how to set the relative `delta` parameters of the different descriptors.

I hope this helps.

**Comment by dmilardovich** [comment]
Hi Albapa, 
 
 This helps a lot, thank you very much! I followed [Erhard et al](https://www.nature.com/articles/s41524-022-00768-w#Sec2) to set the delta (the used exactly the same: 4 for 2B and 0.4 for SOAP). Since it worked for the same material, I assume it should be OK for my case. 

 I have just tried training on energies AND forces. The training dataset contains roughly 1000 configurations (220 - 250 atoms each). Energies and forces were calculated with DFTB. The gap_fit command is next:

os.system("gap_fit energy_parameter_name=energy force_parameter_name=forces at_file=GAP_TRAIN_forces.xyz gap={distance_Nb order=2 cutoff=4.0 n_sparse=30 covariance_type=ard_se delta=4.0 theta_uniform=2.0 sparse_method=uniform compact_clusters=T : soap atom_sigma=0.5 l_max=6 n_max=12 cutoff=4.0 cutoff_transition_width=1.0 delta=0.4 covariance_type=dot_product n_sparse=1000 zeta=4} default_sigma={0.002 0.2 0.0 0.0} config_type_sigma={bulk:0.005:0.05:0.0:0.0:cluster:0.01:0.1:0.0:0.0} e0={Si:-29.7169:O:-83.9795:H:-6.4926} sparse_jitter=1e-8 gp_file=GAP_11.xml")

 As a testing dataset I used one of the trajectories from which the training data was sampled. Therefore, the training and testing data are very correlated. However, the accuracy I find is quite bad: 

![TEST_0](https://user-images.githubusercontent.com/47321981/184134181-8908255f-bf06-44f1-a497-35f36d9041d3.png)
 Note: MAE is in meV/atom. 

 What could be going on here? Is there anything wrong in the gap_fit command? What could I change/try in the fit command? 

 Thanks a lot once again!

**Comment by albapa** [comment]
It's hard to say, but you need to look at the number of sparse points, cutoffs (and corresponding `n_max` parameters in SOAP) etc. If your model is not flexible enough to represent the training data (or there is some inherent problem with your training data), you'll see behaviour like that, when you cannot reproduce the training data as your model can only learn some kind of mean approximation. The other extreme is too much flexibility and little data: that's when you overfit and you need to catch that with appropriate validation.

Best regards
Albert

**Comment by dmilardovich** [comment]
Thanks again Albert! I suspect there is a problem with my data, I will check that and re-train. 
 
 Best, 
 Diego

**Comment by gabor1** [comment]
LEt me chip in here to reinforce Albert's point on forces: there has never been a GAP in any publication doing anything useful that was not trained on forces. I wouldn't trust any potential (even for energy accuracy) that was not trained on forces. Your training command otherwise looks like a good start. I would make the 2b cutoff the same as the soap cutoff. the delta 2b is large, but not catastrophic. one can check the 2b potential itself to see if it ended up being pathological . since you have dimers in the dataset, just plot the dimer curve of the potential (that includes 2b + soap). do you have the isolated atom in the dataset? it's the preferred way of setting the e0-s since having them in the dataset also guarantees that their energies will be right (and I usually make their corresponding energy sigma very small, like 10^-4 or 10^-5)

**Comment by gabor1** [comment]
Also, the recommendation for l_max and n_max has been n_max = 2*lmax for a while, you are using the old heuristic of n_max=l_max. so switch to n_max=8 l_max=4

**Comment by gabor1** [comment]
also, you should have atom_sigma*n_max ~ cutoff so that's another reason to increase your n_max

**Comment by gabor1** [comment]
your energy sigmas are quite high for the bulk (amorphous and cluster are ok). you could easily go down to 0.002 or 0.001, once you add forces (with force sigma 0.01-0.02)

**Comment by dmilardovich** [comment]
Thanks Albert and Gabor once again! I am already running new trainings/testings following your recommendations. I will post the results when they are ready.

**Comment by dmilardovich** [comment]
Hi all, thank you very much once again for your input. The GAP seems to be working very well now. The main issue was that the forces were in the wrong units.
